### PR TITLE
✨ Make the web picker usable on a touch screen

### PR DESCRIPTION
**Pull request by zeenix**
Two things made the hosted picker close to unusable on a phone: tapping the
search box did nothing, so the on-screen keyboard never appeared and there was
no way to type; and the list could only be moved with arrow keys, so dragging
it did nothing.

**This is a breaking release for `gimoji-core`** — see the release note at the
bottom.

## The keyboard

The picker paints into a `<canvas>`, which can't take focus and so can't raise
a keyboard on its own — there was nothing for the browser to bring one up for.
An offscreen `<input>` now stands in for it:

* Tapping the search box focuses it, from the pointer handler itself, since
  browsers only raise the keyboard for a focus change a user gesture drove. It
  blurs first when the element already holds focus, because `focus()` on the
  focused element is a no-op — which would otherwise strand anyone who
  dismissed the keyboard with its own "done" key. For the same reason the
  startup focus (so a desktop visitor can type immediately) is skipped on
  coarse pointers.
* Its `input` events, not `keydown`, are what the search text is built from.
  Mobile keyboards routinely report characters through `input` alone and
  rewrite whole spans via autocorrect, swipe typing and IME composition, so the
  new `Action::SetSearch` replaces the text wholesale rather than replaying
  keys. Physical keyboards keep working either way.
* `keydown` handles navigation only, and bows out entirely while an IME is
  composing — those presses drive the candidate window, so claiming Enter or
  Escape there would hijack the composition rather than support it.
* The press suppresses its compatibility mouse events. Those arrive *after*
  `pointerup` and their default action moves focus to `<body>`, which was
  silently undoing the `focus()` call and taking the keyboard back down.

Raising the keyboard also shrinks the viewport, and can pan it, so the page now
sizes *and positions* itself against `visualViewport` instead of `100vh`:
Android shrinks the layout viewport too, but iOS does neither, and the list
would sit behind the keyboard or off the side of it.

## The scrolling

A press is now resolved on release, so a drag that moved more than a few pixels
scrolls the list instead of picking the emoji out from under the finger that
was scrolling it. A wheel handler shares the same path on desktop.
`touch-action: none` is what stops the browser claiming the drag for panning or
double-tap zoom first.

The supporting `Action::Scroll` clamps at both ends rather than wrapping, and
carries the selection along — the table widget scrolls itself back to whatever
is selected on the next render, which would otherwise undo each scroll
immediately.

## Selection follows the pointer

Clicking a row copied its emoji but left the highlight wherever the keyboard
had last put it, usually the top of the list; the next arrow key jumped back
there and a following Enter picked something the user had never pointed at.
`Action::PickAt` now moves the selection onto the picked row, so pointer and
keyboard share one cursor. This is device-independent — it was just as wrong
with a mouse. Nothing changes for the native CLI, which has no pointer input.

## Release note

`Action` is public and was exhaustively matchable, so the new variants break
downstream matches — hence the `💥` commit and the major bump. `Action` is now
`#[non_exhaustive]` so future additions don't repeat this.

Making that actually happen needed the release config: `💥` was listed in
`custom_minor_increment_regex` with every other feature emoji and there was no
major regex at all, so a breaking-change commit shipped as a minor release with
`semver_check` off to catch it. Also dropped `gimoji-web`'s version requirement
on `gimoji-core`: that package is excluded from the root workspace, so
release-plz never rewrites it, and a core 2.0.0 leaves the wasm build unable to
resolve its own path dependency (verified locally by simulating the bump).

## Testing

`gimoji-core` gets unit tests for the new actions and for the selection
behaviour. The frontend was driven in Chromium at 390×844 with touch emulation:
no pre-focus on a coarse pointer, a tap focuses, a re-tap emits a real
blur→focus pair, typing filters, an autocorrect-style value rewrite filters
correctly, Escape clears both picker and element, a touch drag scrolls and
clamps, a tap picks and moves the highlight, a *drag* does not pick, and
desktop wheel and typing still work. IME behaviour was checked by driving a
real composition through `Input.imeSetComposition`: Enter and Escape
mid-composition no longer reach the picker, and work again once it ends.

Two things emulation can't establish, both worth a real-device pass before
merging: whether iOS Safari and Android Chrome actually raise the keyboard on
each focus transition (load → tap → dismiss → tap again), and viewport panning
while the keyboard is open.

The web frontend has no test harness — `crates/gimoji-web` is pinned to
`wasm32-unknown-unknown` and `just test` only runs the native workspace — so
the browser checks above are scripted but not part of CI. Happy to add
`wasm-bindgen-test` plus a headless runner as a follow-up if that coverage is
wanted.

## Not fixed here

Faint ghost pixels remain on the list's top padding row when content scrolls —
glyph ascenders bleed one row up into a cell ratatui never marks dirty, so the
canvas backend never repaints it. This predates the change (it reproduces with
plain arrow-key scrolling on `main`); drag scrolling just makes it happen more
often. Fixing it means work in `canvas_backend.rs`, left alone as out of scope.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_015ndXLfPQQJYRTReLiTvDpx

### CODE REVIEW

File: crates/gimoji-core/src/app.rs [hunk 1]
```diff
@@ -17,9 +17,15 @@ use crate::{
 pub enum Action {
     Append(char),
     Backspace,
+    /// Replace the search text wholesale — see [`SearchEntry::set_text`].
+    SetSearch(String),
```
Comment (zeenix) [inline crates/gimoji-core/src/app.rs:28]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

**Blocking:** `Action` is a public, exhaustively matchable enum re-exported by the already-released `gimoji-core` 1.4.0, so adding `SetSearch` and `Scroll` breaks downstream exhaustive matches. This `✨` commit is configured to trigger a minor bump, and `semver_check` is disabled. Please either avoid extending this enum or ship this as a major breaking release; with a major bump, making `Action` non-exhaustive would prevent the same problem on future additions.

File: crates/gimoji-web/src/lib.rs [hunk 2]
```diff
@@ -60,20 +87,33 @@ pub fn run() -> Result<(), JsValue> {
...
     install_tick(&window, &state);
     install_resize(&window, &terminal);
+    install_viewport_sync(&window, &document, &terminal);
     install_raf_loop(&window, &terminal, &state);
 
+    // Take focus up front so a desktop visitor can type straight away. On
+    // mobile this doesn't raise the on-screen keyboard — browsers only do
+    // that for a focus change driven by a user gesture — which is what the
+    // tap-on-the-search-box path in `install_pointer_gestures` is for.
+    state.borrow().text_input.focus();
```
Comment (zeenix) [inline crates/gimoji-web/src/lib.rs]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

**Blocking:** This focuses the input during startup, outside user activation. Mobile browsers can leave it as `document.activeElement` without raising the keyboard; the later search tap then calls `focus()` on an already-focused element, so the user gesture causes no focus transition and may still not raise the keyboard. Avoid pre-focusing it on coarse/touch devices (or synchronously blur/refocus), and verify the exact load → tap sequence on real Android and iOS devices.

File: crates/gimoji-web/src/input.rs [hunk 3]
```diff
@@ -6,7 +6,18 @@ use web_sys::KeyboardEvent;
...
+/// `crate::text_input`) currently has focus. When it does, editing keys are
+/// left alone: the element applies them itself and reports the result as an
+/// `input` event, which becomes an [`Action::SetSearch`]. Claiming them here
+/// too would either double-apply the edit or, via `preventDefault`, stop the
+/// element from ever seeing it.
+pub fn from_keyboard(
+    event: &KeyboardEvent,
+    search_is_empty: bool,
+    text_input_focused: bool,
+) -> Option<Action> {
```
Comment (zeenix) [inline crates/gimoji-web/src/input.rs:20]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

**Blocking:** Browsers continue sending `keydown` events during IME composition with `isComposing = true`. This handler still claims Enter, Escape, ArrowUp, and ArrowDown, and the caller prevents their defaults, so it can pick/clear/move instead of letting the IME confirm, cancel, or navigate candidates—the opposite of the IME support intended here. Return `None` for composing events and add a composition regression test.

File: crates/gimoji-web/src/lib.rs [hunk 4]
```diff
@@ -301,20 +562,89 @@ fn install_color_scheme_listener(state: &Rc<RefCell<State>>) {
...
+            .add_event_listener_with_callback(name, cb.as_ref().unchecked_ref())
+            .expect("visual viewport listener install");
+    }
+    cb.forget();
+}
+
+/// Publish the visual viewport's height as `--app-height`, returning
+/// whether that changed anything.
+fn publish_viewport_height(root: &HtmlElement, viewport: &VisualViewport) -> bool {
+    let height = viewport.height();
```
Comment (zeenix) [inline crates/gimoji-web/src/lib.rs:652]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

**Blocking for the iOS keyboard path:** the visual viewport can pan as well as resize; `offsetTop`/`offsetLeft` changes are what its `scroll` event reports. This callback subscribes to `scroll` but publishes only the height, and returns `false` when only the offset changes. When Safari pans to the focused hidden input, the body/canvas remains anchored at the layout-viewport origin and can be clipped or misaligned. Publish/apply the offsets too (or otherwise pin the canvas to the visual viewport), then cover keyboard-open viewport panning on a real iOS device.

**Review by zeenix — COMMENTED** [review]
🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

CI is green, `main` has not moved, and the three commits are self-contained and correctly ordered per `CONTRIBUTING.md`. Four blocking findings are inline; the browser-specific ones need real-device coverage because Chromium touch emulation cannot exercise the virtual keyboard and visual-viewport behavior.

File: crates/gimoji-core/src/app.rs (hunk 1 above)
Comment (zeenix) [reply crates/gimoji-core/src/app.rs:28]: Agreed, and taken the major-release route with the `#[non_exhaustive]` you suggested.

`Action` is now `#[non_exhaustive]`, and the commit adding the variants is retitled `💥` so the breakage is announced rather than implied.

That alone wasn't enough to make it a major release, though: `custom_minor_increment_regex` listed `💥` alongside every other feature emoji and there was no major regex at all, so a `💥` commit bumped the minor version like anything else — with `semver_check` off, nothing would have caught it. Added `custom_major_increment_regex = "^💥"` and dropped `💥` from the minor list so the two can't both claim a commit.

One thing that fell out of testing the bump: `crates/gimoji-web` is excluded from the root workspace, so release-plz never rewrites its requirement on `gimoji-core`. I simulated the bump locally by setting the core version to 2.0.0, and the wasm build stops resolving its own path dependency:

```
error: failed to select a version for the requirement `gimoji-core = "^1.4.0"`
candidate versions found which didn't match: 2.0.0
```

Since that package sets `publish = false`, the requirement bought nothing, so it's now a bare path dependency. Both changes are in `🔧 Release breaking-change commits as a major bump`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-web/src/lib.rs (hunk 2 above)
Comment (zeenix) [reply crates/gimoji-web/src/lib.rs]: Good catch — this was a real dead end, and I've done both things you suggested.

1. Startup no longer pre-focuses when `(pointer: coarse)` matches, so a touch device reaches the first tap with focus still on `<body>` and the tap is a genuine transition. Desktop keeps the pre-focus and can type immediately.
2. The tap path now blurs first when the element already holds focus (`TextInput::focus_from_gesture`). That covers the case the coarse-pointer check can't: someone dismisses the keyboard with its own "done" key, focus stays on the input, and a plain `focus()` would be a no-op forever after.

Verified in Chromium with touch emulation (390×844, `pointer: coarse`):

```
coarse pointer matches: true
active on load (want (body)): (body)
active after search tap (want gimoji-input): gimoji-input
focus events on re-tap (want blur,focus): blur,focus
desktop pre-focus (want gimoji-input): gimoji-input
```

The last two lines are the parts that were broken before: the re-tap now emits a real `blur`→`focus` pair rather than nothing.

On your point about real devices — agreed, and I can't do that from here. Emulation confirms the focus *transitions* happen, which is the mechanism, but whether iOS Safari and Android Chrome actually raise the keyboard on each of them is exactly what emulation can't tell us. Worth walking through load → tap → dismiss → tap again on both before merging.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-web/src/input.rs (hunk 3 above)
Comment (zeenix) [reply crates/gimoji-web/src/input.rs:20]: You're right, and the irony wasn't lost on me — the commit message cited IME composition as a reason for `SetSearch` while the key handler was hijacking the composition. `from_keyboard` now returns `None` outright when `event.is_composing()`, before any other branch, so Ctrl-C is left to the IME too.

On the regression test: I can't add a runnable one without new infrastructure, so I want to flag that rather than quietly skip it. `from_keyboard` takes a `web_sys::KeyboardEvent`, which can't be constructed off-target, and `crates/gimoji-web/.cargo/config.toml` pins the package to `wasm32-unknown-unknown` — so any `#[cfg(test)]` there needs `wasm-bindgen-test` plus a headless browser runner in CI, none of which exists today. `just test` only runs the native workspace. I didn't want to bolt a browser test harness onto this PR uninvited; happy to do it as a follow-up if you'd like the coverage.

What I did instead was drive a real composition through Chromium's DevTools Protocol (`Input.imeSetComposition`, then `dispatchKeyEvent` with `isComposing: true`):

```
composing text in the element: "にほん"
clipboard after Enter while composing (want SENTINEL): "SENTINEL"
element value after Escape while composing (want kept): "にほん"
clipboard after Enter with no composition (want an emoji): "🎨"
```

So mid-composition Enter no longer picks and Escape no longer clears, and both work normally again once composition ends. Before the fix the first of those returned an emoji instead of `SENTINEL`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-web/src/lib.rs (hunk 4 above)
Comment (zeenix) [reply crates/gimoji-web/src/lib.rs:652]: Correct — I was subscribing to `scroll` and then ignoring everything it reports. Fixed by pinning the page to the visual viewport rather than only sizing to it.

`publish_viewport_height` is now `publish_viewport_metrics` and publishes `--app-offset-left` / `--app-offset-top` alongside `--app-height`; `body` carries `transform: translate(var(--app-offset-left, 0px), var(--app-offset-top, 0px))`, so a pan moves the picker with the visual viewport instead of stranding it at the layout-viewport origin.

The return value now means "the height changed", which is deliberate: offsets only translate the page, while a height change resizes the canvas and needs the grid re-measured. That keeps the repaint off the pan/pinch path, which fires continuously, while still repainting when the keyboard opens.

Verified in Chromium by driving the variables directly (`visualViewport` offsets aren't settable from a test):

```
body transform at rest: matrix(1, 0, 0, 1, 0, 0)
body transform after a simulated pan: matrix(1, 0, 0, 1, 20, 120)
canvas moved with it: top 62 -> 182
```

Two caveats I'd rather state than paper over. The transform makes `body` the containing block for the fixed-position offscreen input — harmless here since `body` fills the viewport, but it's a real consequence. And I've left `visualViewport.scale` alone: compensating for pinch-zoom is a different problem from the keyboard one this fixes, and `touch-action: none` blocks pinch over the canvas anyway. Same caveat as the other threads on the real-device side — panning while the keyboard is open on iOS is exactly what emulation can't reproduce.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-web/src/input.rs (hunk 3 above)
Comment (zeenix) [reply crates/gimoji-web/src/input.rs:20]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

This fixes steady-state composition, but not its boundary keydowns. `compositionstart` may fire after the opening `keydown`, and `compositionend` before the closing one, so `is_composing()` is false even though the IME still marks that event with `keyCode == 229`. [MDN recommends checking both signals](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#keydown_events_with_ime). Please also return `None` for `event.key_code() == 229` (or track composition state) and exercise the opening/closing boundaries; otherwise an edge Enter, Escape, or arrow can still reach the picker.

File: crates/gimoji-web/Cargo.toml [hunk 5]
```diff
@@ -43,20 +47,25 @@ web-sys = { version = "0.3.72", features = [
     "Document",
     "HtmlElement",
     "HtmlCanvasElement",
+    "HtmlInputElement",
```
Comment (zeenix) [inline crates/gimoji-web/Cargo.toml:50]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

**Blocking on commit hygiene:** all five new `web-sys` gates (`HtmlInputElement`, `CssStyleDeclaration`, `WheelEvent`, `AddEventListenerOptions`, and `VisualViewport`) currently live in `cbc9333` (`🔧 Release breaking-change commits as a major bump`), while the code that requires them arrives in `b93f85c`. That leaves unrelated touch-feature prerequisites hidden in the release-plumbing commit. Per `CONTRIBUTING.md`’s atomic-commit rule, please move these gates into `b93f85c`, keeping only the core path-dependency and release-plz changes in `cbc9333`.

File: crates/gimoji-core/src/app.rs [hunk 6]
```diff
@@ -13,13 +13,25 @@ use crate::{
     toast::Toast,
 };
 
+/// Something the user did, for [`App::handle`] to interpret.
+///
+/// Marked `#[non_exhaustive]`: new input sources keep needing new actions —
+/// this release alone added three — and each one would otherwise break every
```
Comment (zeenix) [inline crates/gimoji-core/src/app.rs]: 🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

This release adds two `Action` variants (`SetSearch` and `Scroll`), not three; `PickAt` was already present in `gimoji-core` 1.4.0. Please correct this public API documentation so its rationale stays factual.

**Review by zeenix — COMMENTED** [review]
🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

Recheck of `402bced`: both workflows pass and `main` is unchanged. The SemVer, startup-focus, and viewport code paths now address their original findings; the real-device mobile pass called out in the PR is still needed. The IME boundary blocker remains in its thread, with two additional findings inline.

File: crates/gimoji-web/src/input.rs (hunk 3 above)
Comment (zeenix) [reply crates/gimoji-web/src/input.rs:20]: Right again — `is_composing()` alone left both edges open. `from_keyboard` now bails on `event.is_composing() || event.key_code() == 229`.

I checked this was a real gap rather than one already covered by accident, since a boundary keydown reported as `key: "Process"` would have fallen through the `match` regardless. The case that bites is the one where `key` is the real key name with `keyCode 229`, so that's what I exercised — against a build *without* the new clause:

```
Enter,  keyCode 229, isComposing=false -> clipboard: "🐛"    (picked!)
Escape, keyCode 229                    -> search text: ""     (cleared!)
arrows, keyCode 229                    -> then Enter picked "🔥" (selection moved!)
```

All three now behave:

```
Enter,  keyCode 229, isComposing=false -> clipboard: "SENTINEL"
Escape, keyCode 229                    -> search text: "fix"
arrows, keyCode 229                    -> then Enter picked "🐛"
```

Steady-state composition still bails (`Enter` with `isComposing=true` leaves the clipboard alone), and ordinary keys are unaffected — plain Enter picks `🐛`, ArrowDown then Enter picks `🚑️` — so the check isn't swallowing real input. No genuine key reports `keyCode` 229, which is what makes it safe to treat as "not ours".

I went with the `keyCode` check rather than tracking composition state via `compositionstart`/`compositionend`, since the state-tracking approach has the same edge problem in reverse: the flag would have to be set before the events that define it. Still no runnable regression test for the reasons in my earlier reply — this is scripted through CDP, not CI.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-web/Cargo.toml (hunk 5 above)
Comment (zeenix) [reply crates/gimoji-web/Cargo.toml:50]: Fair — I staged the whole file into the release commit because it carried the path-dependency change, and swept the five gates along with it. Rewritten so `634d3f3` touches only `release-plz.toml` and the `gimoji-core` requirement, and all five gates land in the touch-feature commit alongside the code that needs them:

```
$ git show 634d3f3 -- crates/gimoji-web/Cargo.toml   # release plumbing
+gimoji-core = { path = "../gimoji-core" }            # (plus its comment)

$ git show 6c1e9f7 -- crates/gimoji-web/Cargo.toml   # touch features
+    "HtmlInputElement",
+    "CssStyleDeclaration",
+    "WheelEvent",
+    "AddEventListenerOptions",
+    "VisualViewport",
```

Since the split moves build inputs between commits, I checked every commit still builds standalone rather than assuming it — `cargo check` on both cargo roots at each of the five, in a detached worktree, all clean. The release commit compiles without the gates because the web sources at that point are still `main`'s and don't reference those APIs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/gimoji-core/src/app.rs (hunk 6 above)
Comment (zeenix) [reply crates/gimoji-core/src/app.rs]: Correct, my mistake — `PickAt` shipped in 1.4.0 with the WebAssembly frontend. I was counting the three additions in the commit as a whole (`Scroll`, `SetSearch`, and `App::search_area`), but `search_area` is a method, not a variant, so "three actions" was wrong wherever it appeared.

The doc comment now reads "this release adds two of them for pointer and touch input", and I fixed the same miscount in the commit message body rather than leave the two disagreeing.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by zeenix — COMMENTED** [review]
🤖 ChatGPT GPT-5.6 Sol, reviewing on behalf of Zeeshan.

Rechecked `f126f71`. The IME boundary guard, commit split, and public API documentation now address all remaining review threads; those threads are resolved. Both workflows pass, `main` has not moved, and `git diff --check` is clean. No further code findings.

The real iOS/Android focus sequence and iOS keyboard-open viewport pass already documented in the PR remain the only pre-merge verification.
